exonum/src/messages2/protocol.rs

```diff
@@ -0,0 +1,709 @@
+    ///
+    /// ### Generation
+    /// This message can be sent during `Status` processing.
+
+    struct BlockRequest {
+        /// Public key of the recipient.
+        to: & PublicKey,
+        /// The height to which the message is related.
+        height: Height,
+    }
+}
+
+impl BlockResponse {
+    /// Verify Merkle root of transactions in the block.
+    pub fn verify_tx_hash(&self) -> bool {
+        let db = MemoryDB::new();
+        let mut fork = db.fork();
+        let mut index = ProofListIndex::new("verify_tx_hash", &mut fork);
+        index.extend(self.transactions().iter().cloned());
+        let tx_hashes = index.merkle_root();
+        tx_hashes == *self.block().tx_hash()
+    }
+}
+
+impl Precommit {
+    /// Verify precommits signature and return it's safer wrapper
+    pub(crate) fn verify_precommit(
+        buffer: Vec<u8>,
+    ) -> Result<Message<Precommit>, ::failure::Error> {
+        let signed = SignedMessage::verify_buffer(buffer)?;
+        let protocol = Protocol::deserialize(signed)?;
+        ProtocolMessage::try_from(protocol)
+            .map_err(|_| format_err!("Couldn't verify precommit from message"))
+    }
+}
+
+/// Full message constraints list.
+#[doc(hidden)]
+pub trait ProtocolMessage: Debug + Clone + BinaryForm {
+    fn message_type() -> (u8, u8);
+    ///Trying to convert `Protocol` to concrete message,
+    ///if ok returns message `Message<Self>` if fails, returns `Protocol` back.
+    fn try_from(p: Protocol) -> Result<Message<Self>, Protocol>;
+
+    fn into_protocol(this: Message<Self>) -> Protocol;
+
+    fn into_message_from_parts(self, sm: SignedMessage) -> Message<Self>;
+}
+
+/// Implement Exonum message protocol.
+///
+/// Protocol should be described according to format:
+/// ```
+/// /// type of SignedMessage => new name of Protocol enum.
+/// SignedMessage => Protocol {
+///       // class ID => class name
+///       0 => Service {
+///            // message = message type ID
+///            RawTransaction = 0,
+///            Connect = 1,
+///            Status = 2,
+///            // ...
+///        },п
+///        1 => Consensus {
+///            Precommit = 0,
+///            Propose = 1,
+///            Prevote = 2,
+///        },
+/// }
+/// ```
+/// where:
+/// `$ProtocolName` is an name of protocol which is described by this schema,
+///     enum with same name will be created. All message classes are encapsulate by this enum;
+/// `$MessageClass` is a module name which is designed to handle messages,
+///     enum with same name will be crated. All message within `MessageClass` should be unique;
+/// `$MessageType` is a concrete messages within some `$MessageClass`;
+/// `$SignedMessage` is a typename of `SignedMessage` from which $ProtocolName could be created;
+/// `$cls_num` and `$typ_num` is a constant which represent
+///     message class and message type respectively
+///
+/// Each `$MessageType` should implement `Clone` and `Debug`.
```

Comment:
These `$names` now do not refer to anything. I guess this should be rewritten to instruct the users in plain prose.